diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Scan.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Scan.java
index 708d6c7356..0be51c41ef 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Scan.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Scan.java
@@ -895,7 +895,9 @@ public class Scan extends Query {
   @Deprecated
   public Scan setSmall(boolean small) {
     this.small = small;
-    this.readType = ReadType.PREAD;
+    if (small) {
+      this.readType = ReadType.PREAD;
+    }
     return this;
   }
 
